src/Data/Parameterized/MonoidSolver.hs

```diff
@@ -0,0 +1,210 @@
+{-# LANGUAGE UndecidableInstances #-}
+
+module Data.Parameterized.MonoidSolver
+  ( solve
+  , solveNat
+  ) where
+
+import Data.Kind (Type)
+import Data.Proxy (Proxy(..))
+import Data.Type.Equality (type (:~:)(Refl))
+import GHC.TypeLits (Nat)
+
+import Data.Parameterized.NatRepr (NatRepr, type (+), plusAssoc)
+
+-- | 'Op' corresponds to @Add@ from the paper.
+type family Op k (m1 :: k) (m2 :: k) :: k
+-- | 'Unit' corresponds to @Add@ from the paper.
+type family Unit k :: k
+
+type instance Op Nat n1 n2 = n1 + n2
+type instance Unit Nat = 0
+
+-- | Type level only
+data MonoidExpr k
+  = EUnit
+  | EVar k
+  | EOp (MonoidExpr k) (MonoidExpr k)
+
+data MERepr (k :: Type) (f :: k -> Type) (me :: MonoidExpr k) where
+  EUnitRepr :: MERepr k f 'EUnit
+  EVarRepr :: f n -> MERepr k f ('EVar n)
+  EOpRepr ::
+    MERepr k f me1 ->
+    MERepr k f me2 ->
+    MERepr k f ('EOp me1 me2)
+
+type family Eval (k :: Type) (me :: MonoidExpr k) :: k where
+  Eval k 'EUnit = Unit k
+  Eval k ('EVar m) = m
+  Eval k ('EOp me1 me2) = Op k (Eval k me1) (Eval k me2)
+
+-- | Defunctionalization - inspired by \"singletons\", and with a similar naming
+-- scheme
+data TyFun a b
+type a ~> b = TyFun a b -> Type
+type family Apply (f :: a ~> b) (x :: a) :: b
+
+data IdSym1 :: MonoidExpr m ~> MonoidExpr m
+data EOpSym0 :: MonoidExpr m ~> (MonoidExpr m ~> MonoidExpr m)
+data EOpSym1 (me :: MonoidExpr m) :: MonoidExpr m ~> MonoidExpr m
+data ComposeDiffSym0 :: MonoidExpr m ~> (MonoidExpr m ~> (MonoidExpr m ~> MonoidExpr m))
+data ComposeDiffSym1 (me :: MonoidExpr m) :: MonoidExpr m ~> (MonoidExpr m ~> MonoidExpr m)
+data ComposeDiffSym2 (me1 :: MonoidExpr m) (me2 :: MonoidExpr m) :: MonoidExpr m ~> MonoidExpr m
+
+type instance Apply IdSym1 me = me
+type instance Apply EOpSym0 me = EOpSym1 me
+type instance Apply (EOpSym1 me1) (me2 :: MonoidExpr m) = 'EOp me1 me2
+-- TODO: generalize to (higher-order) compose?
+type instance Apply ComposeDiffSym0 me = ComposeDiffSym1 me
+type instance Apply (ComposeDiffSym1 me1) me2 = ComposeDiffSym2 me1 me2
+type instance Apply (ComposeDiffSym2 me1 me2) me3 = Apply (Diff me1) (Apply (Diff me2) me3)
+
+-- | The \"difference list\"/Caley embedding representation of monoid
+-- expressions, corresponds to the bracket operator in the paper.
+type family Diff (me :: MonoidExpr m) :: MonoidExpr m ~> MonoidExpr m where
+  Diff ('EOp me1 me2) = ComposeDiffSym2 me1 me2
+  Diff 'EUnit = IdSym1
+  Diff ('EVar m) = EOpSym1 ('EVar m)
+
+-- | 'UnDiff' corresponds to "reify" from the paper.
+type family UnDiff (diff :: MonoidExpr m ~> MonoidExpr m) :: MonoidExpr m where
+  UnDiff diff = Apply diff 'EUnit
+
+type family Normalize (me :: MonoidExpr m) :: MonoidExpr m where
+  Normalize me = UnDiff (Diff me)
+
+-- | Note on termination: The one recursive call to this function is strictly
+-- decreasing on its first argument. This function is mutually recursive with
+-- 'normalizeSound' and recursive calls to that function are made on the second
```

Comment:
Again, I'm not sure if "second argument" is strictly accurate here.

Reply:
Like the above, this is fixed by packaging the axioms into a typeclass.